core/lib/zksync_core/src/api_server/web3/pubsub.rs

```diff
@@ -238,118 +238,174 @@ impl EthSubscribe {
         self.events_sender = Some(sender);
     }
 
-    /// Assigns ID for the subscriber if the connection is open, returns error otherwise.
-    fn assign_id(
-        subscriber: typed::Subscriber<PubSubResult>,
-    ) -> Result<(typed::Sink<PubSubResult>, SubscriptionId), ()> {
-        let id = H128::random();
-        let sub_id = SubscriptionId::String(format!("0x{}", hex::encode(id.0)));
-        let sink = subscriber.assign_id(sub_id.clone())?;
-        Ok((sink, sub_id))
+    async fn reject(sink: PendingSubscriptionSink) {
+        sink.reject(ErrorObject::borrowed(
+            ErrorCode::InvalidParams.code(),
+            "Rejecting subscription - invalid parameters provided.",
+            None,
+        ))
+        .await;
+    }
+
+    async fn heads_subscriber(
+        pending_sink: PendingSubscriptionSink,
+        b: broadcast::Receiver<BlockHeader>,
```

Comment:
I'd suggest to use `PubSubResult` as the channel element type so that this logic can be reused for blocks and transactions.

Reply:
done